src/directives/tooltip.js

```diff
@@ -0,0 +1,329 @@
+  const tooltip = tooltips.get(event.target);
+  if (!tooltip.shown) tooltip.show();
+};
+const hide = (event) => {
+  const tooltip = tooltips.get(event.target);
+  if (tooltip.shown) tooltip.hide();
+};
+
+class Tooltip {
+  static maybeCreate(anchor, { modifiers, value }) {
+    const type = types.find(name => modifiers[name] === true);
+    if (type == null) {
+      // eslint-disable-next-line no-console
+      console.error('invalid tooltip', anchor);
+      return null;
+    }
+
+    const tooltip = new Tooltip(anchor, type, value);
+    /*
+    For v-tooltip.text, we check whether the text is more than 5 characters
+    rather than checking whether tooltip.text() != null because the element that
+    sets text-overflow might not be visible yet. In that case, clientWidth === 0
+    on that element, so tooltip.text() == null even if the text will be
+    truncated once the element is visible. Rather than trying to watch the
+    visibility of the element, we add event listeners if the text is more than 5
+    characters. If/when an event listener is triggered for show, we will
+    evaluate then whether the text is truncated and show the tooltip if so.
+    Presumably the text won't ever be truncated if it is 5 characters or fewer.
+
+    For v-tooltip.sr-only, we always add event listeners because we can't easily
+    hook into the lifecycle of the .sr-only element. If/when an event listener
+    is triggered for show, we will evaluate then whether there is an .sr-only
+    element and whether it has text and show the tooltip if so.
+
+    For other tooltip types, we can know right away whether a tooltip should be
+    shown: we just need to check tooltip.text().
+    */
+    if (type === 'text'
+      ? anchor.textContent.trim().length > 5
+      : type === 'sr-only' || tooltip.text() != null) {
+      tooltip.addEventListeners();
+      if (type === 'aria-describedby') tooltip.render();
+      tooltips.set(anchor, tooltip);
+      return tooltip;
+    }
+
+    return null;
+  }
+
+  constructor(anchor, type, value) {
+    this.anchor = anchor;
+    // I tried subclassing based on `type`, but it didn't end up working well:
+    // `type` is used in a wide variety of ways.
+    this.type = type;
+    this.value = value;
+    this.element = null;
+    this.shown = false;
+  }
+
+  _text() {
+    if (this.type === 'text') {
+      let closestBlock = this.anchor;
+      // clientWidth === 0 if the element is inline.
+      while (closestBlock.clientWidth === 0) {
+        closestBlock = closestBlock.parentNode;
+        // The `null` case is possible in testing, because many components are
+        // not attached to the document, in which case clientWidth is also 0.
+        // The document.body case shouldn't happen, but it seems like a good
+        // safeguard.
+        if (closestBlock == null || closestBlock === document.body) return null;
+      }
+      return closestBlock.scrollWidth > closestBlock.clientWidth
```

Comment:
is this checking if text is truncated?

Reply:
Yes, that's the intention! Maybe I should leave a comment along those lines? 🤔

Reply:
I've left a comment about this.